Support arbitrary course_ids

bdero opened the pull request:
This fix allows an LMS to specify arbitrary course_ids so that dalite-ng
can work with other LMSs besides edX, like moodle.

### edX test instructions:

1. Setup a devstack.
1. Follow these instructions to setup dalite-ng in the devstack, but be sure to checkout this branch (bdero/remove-opaque-keys) when cloning the dalite-ng repo: https://github.com/open-craft/dalite-ng#setting-up-a-development-server-in-an-open-edx-devstack
1. Follow these instructions to setup a dalite-ng problem on a course: https://github.com/open-craft/dalite-ng#usage-example-configuring-dalite-in-edx-studio
1. Publish the new problem in Studio.
1. With the LMS and dalite-ng LTI apps both running, open a new terminal and `vagrant ssh`.
1. With the new ssh session, tail the student log for dalite-ng: `tail -f /edx/src/dalite-ng/log/student.log`
1. Navigate to the dalite-ng LTI problem in the LMS. Observe that there are no errors.
1. Look at the `student.log` tail, notice that the event log just emitted doesn't have an "org_id" field. It should look something like this:

    ```json
    {"username": "QdwEdCbdw1uPCF2LDPVk3g++", "context": {"username": "QdwEdCbdw1uPCF2LDPVk3g++", "course_id": "course-v1:edX+DemoX+Demo_Course", "module": {"usage_key": null}}, "event_source": "server", "event_type": "problem_show", "ip": "192.168.33.1", "agent": "Mozilla/5.0 (X11; Linux x86_64; rv:48.0) Gecko/20100101 Firefox/48.0", "host": "precise64", "referer": "http://localhost:8000/courses/course-v1:edX+DemoX+Demo_Course/xblock/block-v1:edX+DemoX+Demo_Course+type@lti+block@5f53a36d78a34d219c165c3f41fc29d3/handler/preview_handler", "accept_language": "en-US,en;q=0.5", "time": "2016-09-08T20:18:54.578471", "course_id": "course-v1:edX+DemoX+Demo_Course", "event": {"assignment_id": "assign1", "problem": null, "question_text": "What letter does Brandon start with?", "assignment_title": "Test Assignment!", "question_id": 1}}
    ```

    Before the change, it looked like this (notice the "org_id" right after the "course_id" field:
    ```json
    {"username": "QdwEdCbdw1uPCF2LDPVk3g++", "context": {"username": "QdwEdCbdw1uPCF2LDPVk3g++", "course_id": "course-v1:edX+DemoX+Demo_Course", "org_id": "edX", "module": {"usage_key": null}}, "event_source": "server", "event_type": "problem_check", "ip": "192.168.33.1", "agent": "Mozilla/5.0 (X11; Linux x86_64; rv:48.0) Gecko/20100101 Firefox/48.0", "host": "precise64", "referer": "http://192.168.33.10:8321/assignment/assign1/1/", "accept_language": "en-US,en;q=0.5", "time": "2016-09-08T19:40:36.248539", "course_id": "course-v1:edX+DemoX+Demo_Course", "event": {"question_text": "What letter does Brandon start with?", "success": "correct", "assignment_id": "assign1", "assignment_title": "Test Assignment!", "rationale": "sdsdf  ffff", "problem": null, "first_answer_choice": 2, "question_id": 1}}

    ```

### Moodle test instructions:

1. Set up dalite-ng app using the branch from this PR.
1. Configured nginx to use SSL, and proxied dalite-ng.  The proxy configuration was crucial to getting OAuth handshaking to work, but maybe there's another way to do it?

    ```
server {
	listen 443;
	server_name dalite-server-name.com;

	ssl on;
	ssl_certificate /etc/letsencrypt/live/dalite-server-name.com/cert.pem;
	ssl_certificate_key /etc/letsencrypt/live/dalite-server-name.com/privkey.pem;
	ssl_session_timeout 5m;
	ssl_protocols SSLv3 TLSv1;
	ssl_ciphers ALL:!ADH:!EXPORT56:RC4+RSA:+HIGH:+MEDIUM:+LOW:+SSLv3:+EXP;
	ssl_prefer_server_ciphers on;

	location / {
            proxy_pass http://127.0.0.1:8321;
            proxy_redirect http://127.0.0.1:8321 https://dalite-server-name.com;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
	}
}
```
1. Created a couple of Questions and an Assignment as per [README](https://github.com/open-craft/dalite-ng#usage-example-configuring-dalite-in-edx-studio).  Also had to Preview each question, and provide sample "expert" rationales for each possible answer choice, otherwise the student answers couldn't be graded.
1. Set up an individual `External Tool` assignment in Moodle for each question ([Q1](https://moodle.dawsoncollege.qc.ca/course/mod.php?sesskey=gCcZK5Lafb&sr=0&update=15329), [Q2](https://moodle.dawsoncollege.qc.ca/course/mod.php?sesskey=gCcZK5Lafb&sr=0&update=15334)).
1. Impersonated a Student and answered Q1 correctly, and Q2 incorrectly.
1. Checked [Grades](https://moodle.dawsoncollege.qc.ca/grade/report/user/index.php?id=504) to ensure they got recorded ok.

smarnach commented:
@bdero The log format was requested by Harvard, and they have some pipeline running to process the logs.  I don't know whether they are actually using the `org_id` field, but it was in their format specification, which is why I included it.  In any case, I don't want to break their pipeline, so I ask you to catch the exception instead.

Have you tested whether the modified version works on Moodle?  There might be further issues than just this single one, in particular with grading.

bdero commented:
@smarnach Thanks for the review! I've addressed your comments and I think this is ready for another round. I also added a test case for the new code path.

 > In any case, I don't want to break their pipeline, so I ask you to catch the exception instead.

Cool, I switched up this change to make have it included in the logs if the course key is valid. Let me know what you think.

> Have you tested whether the modified version works on Moodle? There might be further issues than just this single one, in particular with grading.

I went ahead and found a way to test this on Moodle, and the grading seems to be working. Moodle sends the correct callback URL and it's getting called appropriately. Unfortunately on Moodle, aren't in the student grade listing, and I only have access to one teacher account on the Dawson Moodle I'm working against.

@bradenmacdonald is there any way I can get access to a student account on Dawson's Moodle in order to test grading? So far two of the vagrant boxes I've tried for a Moodle setup have failed, but I can get it working if getting access to a student account is difficult. ;)

pomegranited commented on peerinst/views.py:
```diff
@@ -151,6 +158,9 @@ def emit_event(self, name, **data):
             username=self.user_token,
         )
 
+        if edx_org:
```
Nit: Could you please make this check `if edx_org is not None`?

bdero replied:
Sure, switched this to make it more explicit, thanks.

pomegranited commented:
@bdero I've requested a few things, but the code for this PR works fine.

CC @smarnach

1. I was confused about why the test instructions are only about setting up Dalite-ng on edX, when this change is supposed to be tested on Moodle?  It took me a while to figure it out, but I was able to get it working with an https deployment of your code, and verified that both auth and grading work.  In Moodle, you can impersonate a Student by scrolling down to the bottom of your Administration menu and selecting `Switch role to... > Student`.  (CC @bradenmacdonald)
1. Please update the PR description to remove the parts about changing the event log and removing the dependency on opaque-keys, since that has since been reverted.
1. I had to add this to `dalite/settings.py` to get OAuth to work behind the https proxy, otherwise the oauth signature wouldn't be accepted.

    ```
SECURE_PROXY_SSL_HEADER = ('HTTP_X_FORWARDED_PROTO', 'https')
```
1. Please update the documentation to include Moodle configuration instructions and screenshots, including how to specify the assignment and question parameters.  For your reference, I've described what I had to do below, and left my app up so you can see it in Dawson's Moodle.
1. Left one [nit](https://github.com/open-craft/dalite-ng/pull/58/files#r78278713).

*Moodle Test Instructions*

1. Set up dalite-ng app using the branch from this PR.
1. Configured nginx to use SSL, and proxied dalite-ng.  The proxy configuration was crucial to getting OAuth handshaking to work, but maybe there's another way to do it?

    ```
server {
	listen 443;
	server_name dalite-server-name.com;

	ssl on;
	ssl_certificate /etc/letsencrypt/live/dalite-server-name.com/cert.pem;
	ssl_certificate_key /etc/letsencrypt/live/dalite-server-name.com/privkey.pem;
	ssl_session_timeout 5m;
	ssl_protocols SSLv3 TLSv1;
	ssl_ciphers ALL:!ADH:!EXPORT56:RC4+RSA:+HIGH:+MEDIUM:+LOW:+SSLv3:+EXP;
	ssl_prefer_server_ciphers on;

	location / {
            proxy_pass http://127.0.0.1:8321;
            proxy_redirect http://127.0.0.1:8321 https://dalite-server-name.com;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
	}
}
```
1. Created a couple of Questions and an Assignment as per [README](https://github.com/open-craft/dalite-ng#usage-example-configuring-dalite-in-edx-studio).  Also had to Preview each question, and provide sample "expert" rationales for each possible answer choice, otherwise the student answers couldn't be graded.
1. Set up an individual `External Tool` assignment in Moodle for each question ([Q1](https://moodle.dawsoncollege.qc.ca/course/mod.php?sesskey=gCcZK5Lafb&sr=0&update=15329), [Q2](https://moodle.dawsoncollege.qc.ca/course/mod.php?sesskey=gCcZK5Lafb&sr=0&update=15334)).
1. Impersonated a Student and answered Q1 correctly, and Q2 incorrectly.
1. Checked [Grades](https://moodle.dawsoncollege.qc.ca/grade/report/user/index.php?id=504) to ensure they got recorded ok.

bdero commented:
@pomegranited Thanks, I addressed all of your comments except for number 4:

> Please update the documentation to include Moodle configuration instructions and screenshots, including how to specify the assignment and question parameters. For your reference, I've described what I had to do below, and left my app up so you can see it in Dawson's Moodle.

By documentation, do you mean the readme in this repository?

pomegranited commented:
@bdero Yep, please update the README in this repo.

And the only other one was point 3: Add `SECURE_PROXY_SSL_HEADER = ('HTTP_X_FORWARDED_PROTO', 'https')` to settings.py.

bdero commented:
@pomegranited My new documentation in the readme is ready for review!

I was able to get the setup to work without SSL when I tested it out earlier by keeping "Force SSL" unchecked when configuring a new "External tool type". I think this can also be done when only configuring an "External tool" (rather than an "External tool type") by specifying `http` instead of `https` as the protocol for the "Launch URL" (as opposed to the "Secure Launch URL"), but I haven't checked in this particular case.

Let me know if you'd still like me to add in the django header config. My thought is that we should turn this on for production configs by default (if there is a sane production config, but I can only find [this default non-production friendly config](https://github.com/open-craft/dalite-ng/blob/master/dalite/settings.py)), but not burden development by enforcing that a webserver performing TLS termination be present in front of Dalite by default.

pomegranited commented:
:+1:  Thanks for making those changes, @bdero !

On the SSL front, my browser complains if I'm trying to post to a non-https site from within an https site, so it would be good to set that flag on their production config.  But you make a good point; there's no reason to mess with the default, devstack configuration.

- [x] I tested this: Followed the README instructions to test the latest code with Moodle and my devstack edX.
- [x] I read through the code.
- [x] Includes documentation - thanks for including Moodle screenshots!

bdero commented:
@pomegranited Thanks for taking another pass!

> On the SSL front, my browser complains if I'm trying to post to a non-https site from within an https site

Ah, I ran into this problem too. I'm not sure about other browsers, but in Firefox it can be disable temporarily by clicking on the "Secure connection information" box (![](http://i.imgur.com/iWy2Iyk.png)) on any page where mixed content iframe blocks are happening. It's great because I used to fight with Firefox/Chromium all the time passing flags that don't work in most cases like `--disable-web-security`.
